students/Nick_Haury/Lesson05/mailroom3.py

```diff
@@ -0,0 +1,172 @@
+amount; donors sorted by largest donation sum to smallest.
+'''
+
+# create initial donor dictionary
+donors = {
+    "William Gates, III": [653772.32, 12.17],
+    "Mark Zuckerberg": [1663.23, 4300.87, 10432.0],
+    "Jeff Bezos": [877.33],
+    "Paul Allen": [663.23, 43.87, 1.32]
+}
+
+
+def quit():
+    print("Closing mailroom... Goodbye")
+    sys.exit()  # exit the script
+
+
+def main():
+    '''
+    Prompt the user for input.  Current choices are to send a thank you or to
+    create a report to display all the current donors and respective donation
+    totals and averages.  The user can also quit.
+    '''
+
+    # functions for users to choose from
+    choice_dict = {
+                    "1": thank_you,
+                    "2": create_report,
+                    "3": write_letters,
+                    "4": quit
+    }
+
+    print("Welcome to the mailroom program")
+    prompt = "\n".join((
+        "Please choose a function from the below options:",
+        "1 - Send a Thank You to a single donor.",
+        "2 - Create a Report.",
+        "3 - Write letters to all donors to file.",
+        "4 - Quit",
+        ">> "
+        ))
+
+    while True:
+        choice = input(prompt)
+        if choice in choice_dict:
+            choice_dict[choice]()
+        else:
+            print(str(choice) + " is an invalid selection, "
+                  "please make a valid selection \n")
+
+
+def thank_you():
+    '''
+    prompts the user for a full name.  If the user types "list", then print
+    a list of donor names and re-prompt.  If a name is typed that is not
+    in the list, add the name to the list.  If a name is typed that is
+    already in the list, then use it.  Once a name is selected, a
+    donation amount is specified.  Donation amount input by users is then
+    added to the donation history of the name.  Email is then automatically
+    drafted using string formatting thanking the donor for the donation.
+    Return to main prompt after.
+    '''
+
+    print()
+    while True:
+        ty_input = input('Type "list" to display a list of donor names. '
+                         'Otherwise enter a name you would like to add a '
+                         'donation for:\n>>')
+
+        if ty_input.lower() == "list":
+            print_donors()
+        else:
+            try:
+                donation_amount = float(input("Please enter donation amount:"
+                                              "\n>>"))
+            except ValueError:
+                print("Please only enter a single number for donation"
+                      " amount\n")
+            else:
+                if donation_amount > 0:
+                    add_donation(ty_input, donation_amount)
+                    print(
+                        "\n", create_email(ty_input, donation_amount), sep=""
+                        )
+                    break
+                else:
+                    print("Please enter a positive number\n")
```

Comment:
I was thinking this is also a good place for exception handling, but I wasn't sure how to create an error for a non-positive number.  After your suggestion on the other spot for exception handling, I was able to figure out how to make it work here too up in the earlier "try" statement.